### ISSUE TITLE
s3: Support empty "directory" keys

### ISSUE THREAD

**Issue by felixbuenemann** [rclone/rclone#753]
Many tools like the the AWS S3 Console, S3 network drives / fuse filesystems and various other tools create "directories" on S3 by creating zero byte keys that end with a slash to emulate empty directories.

rclone currently can't handle these keys, because it tries to treat them like a file instead of a directory and fails with an error if a local directory with the same name exist.

Instead rclone should detect these keys and treat them like a directory.

The keys are easy to detect, because they always end with a slash and have a content-length of 0 bytes.

Here's an example key listed with the python aws cli:

```
aws s3api head-object --bucket bito-pim-images --key 'images/WebShop/ita_IT/'
{
    "AcceptRanges": "bytes",
    "ContentType": "binary/octet-stream",
    "LastModified": "Fri, 22 Jul 2016 12:32:11 GMT",
    "ContentLength": 0,
    "ETag": "\"d41d8cd98f00b204e9800998ecf8427e\"",
    "Metadata": {}
}
```

> What is your rclone version (eg output from `rclone -V`)

v1.33-DEV

> Which OS you are using and how many bits (eg Windows 7, 64 bit)

macOS 10.12.1

> Which cloud storage system are you using? (eg Google Drive)

s3

> The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

`rclone sync --size-only --delete-after remote:tmp /tmp`

> A log from the command with the `-v` flag (eg output from `rclone -v copy /tmp remote:tmp`)

```
2016/09/30 22:09:55 Local file system at /Volumes/EXT4TBA/images: Waiting for transfers to finish
2016/09/30 22:09:55 images/WebShop/ita_IT/: Failed to copy: open /Volumes/EXT4TBA/images/WebShop/ita_IT: is a directory
2016/09/30 22:09:55 images/WebShop/nld_BE/: Failed to copy: open /Volumes/EXT4TBA/images/WebShop/nld_BE: is a directory
2016/09/30 22:09:55 images/WebShop/nld_NL/: Failed to copy: open /Volumes/EXT4TBA/images/WebShop/nld_NL: is a directory
2016/09/30 22:09:55 images/WebShop/pol_PL/: Failed to copy: open /Volumes/EXT4TBA/images/WebShop/pol_PL: is a directory
2016/09/30 22:09:56 images/WebShop/slo_SK/: Failed to copy: open /Volumes/EXT4TBA/images/WebShop/slo_SK: is a directory
2016/09/30 22:09:56 images/WebShop/spa_ES/: Failed to copy: open /Volumes/EXT4TBA/images/WebShop/spa_ES: is a directory
2016/09/30 22:09:56 images/WebShop/ukrainian_ukraine/: Failed to copy: open /Volumes/EXT4TBA/images/WebShop/ukrainian_ukraine: is a directory
2016/09/30 22:09:56 Local file system at /Volumes/EXT4TBA/images: not deleting files as there were IO errors
2016/09/30 22:09:56 Attempt 3/3 failed with 30 errors and: not deleting files as there were IO errors
2016/09/30 22:09:56 Failed to sync: not deleting files as there were IO errors
```

It's very easy to create a test key by going into the AWS S3 Console and clicking on "Create Folder" to create an empty directory.

**Comment by ncw** [comment]
Nice idea thanks.

If you had a list of applications which use a 0 sized `dir\` object to mean a directory that would be very helpful.
 
Note this is related to #14 which discussed s3fs

**Comment by felixbuenemann** [comment]
I believe every single client that supports creating empty directories on s3 uses this pattern.

Here's the one's that I have personally tested:

* The official [AWS S3 Console](https://console.aws.amazon.com/s3/home)
* [TntDrive](https://tntdrive.com) – Windows Network Drive
* [CloudBerry Explorer for Amazon S3](http://www.cloudberrylab.com/free-amazon-s3-explorer-cloudfront-IAM.aspx) – Windows S3 Client
* [Transmit](https://panic.com/transmit/) – Mac File Transfer Client
* [Cyberduck](https://cyberduck.io) – Mac File Transfer Client
* [Forklift](http://www.binarynights.com/forklift/) – Mac File Transfer Client
* [goofys](https://github.com/kahing/goofys) – FUSE S3 filesystem

Notably all of the above are either GUI clients or network drives.

They all create keys that end with a slash, have a ContentLength of 0 bytes and as such all have the ETag "d41d8cd98f00b204e9800998ecf8427e". The ContentType differs between clients, most se either "binary/octet-stream" or "application/octet-stream", Cyberduck uses "application/x-directory".

Pretty much all of the commandline tools do not handle these "directories" properly and fail to handle them during sync / copy either with warning or errors:

* [s3cmd](https://github.com/s3tools/s3cmd/issues/349)
* [minio client](https://github.com/minio/mc/issues/1716)
* rclone (discussed here and in #14)

The official python awscli simply ignores those keys. It never creates empty directories when syncing from local to s3 and never downloads them when syncing or copying from s3 to local. That's not perfect but a lot better than clients that try to compare the key as a regular file with the local directory and error out because of the different types.

**Comment by felixbuenemann** [comment]
I think there are two possible solutions to this problem:

* Ignore those keys (bugfix)
* Handle them like directories with full support for syncing in both directions (enhancement)

Having directory keys ignored as a quick bugfix would be a great improvement and proper handling could be implemented in a future version.

The ergonomic solution for supporting directories is probably to create them locally if directory style keys are found on the remote, but only create them on the remote if empty local directories are synced to the remote. There is however benefit to create those keys for all the directories synced to the remote, because it allows to store the directories modification time, but of course this requires more API calls, so it might not be worth the trouble. I'm not sure how empty directories are handled in other storage providers, but there could be a `--no-empty-dirs` switch to ignore empty directories.

**Comment by timofonic** [comment]
I wonder what's left to make these "directories" work.

@felixbuenemann Has the situation changed in other clients? Do "directory" keys work in Rclone+Minio server?

**Comment by felixbuenemann** [comment]
I believe the issues in minio client have been fixed.

About minio server: Theoretically the server should be agnostic to the problem, since s3 does not have directories and writing zero byte keys to emulate them is just a hack that clients use to work around that limitation.

So if rclone is confused by those keys with the official s3 servers, it is likely it'll have the same problems with minio, which emulates the s3 api.

**Comment by lbahaddur** [comment]
Is the fixes making the "--delete-empty-src-dirs" option futile..  I see that when a directory and its contents created from S3 browser is moved using the "rclone move <src path> <dest path> --delete-empty-src-dirs"  options,  does not remove the empty source directory from the source after the move ..

**Comment by ncw** [comment]
@lbahaddur `--delete-empty-src-dirs` seems to be broken at the moment #2372

However you are right even when it is fixed the backend as it stands at the moment won't delete those directory markers.

**Comment by jamshid** [comment]
I was going to file a bug that `rclone purge` fails if there is a key named with a trailing slash e.g. "`mysubdir/`", because it does not try to delete that key, but I think that is just part of this issue.

**Comment by acejam** [comment]
@jamshid You are correct. I'm seeing the exact same `purge` issue as you.

**Comment by ncw** [comment]
This is now implemented in v1.63 - see https://rclone.org/s3/#s3-directory-markers
